File: src/store/knowledgeBase.ts
import { defineStore } from 'pinia';
import { invoke } from '@tauri-apps/api/core';
import type { KnowledgeBase, Document } from '../types/knowledgeBase';

interface KnowledgeBaseState {
    knowledgeBases: KnowledgeBase[];
    currentKbId: string | null;
    documents: Document[];
    loading: boolean;
    error: string | null;
}

export const useKnowledgeBaseStore = defineStore('knowledgeBase', {
    state: (): KnowledgeBaseState => ({
        knowledgeBases: [],
        currentKbId: null,
        documents: [],
        loading: false,
        error: null,
    }),

    getters: {
        currentKnowledgeBase: (state): KnowledgeBase | undefined => {
            return state.knowledgeBases.find(kb => kb.id === state.currentKbId);
        },
    },

    actions: {
        /** 加载所有知识库 */
        async loadKnowledgeBases() {
            this.loading = true;
            this.error = null;
            try {
                this.knowledgeBases = await invoke<KnowledgeBase[]>('list_knowledge_bases');
            } catch (e) {
                this.error = String(e);
                console.error('加载知识库失败:', e);
            } finally {
                this.loading = false;
            }
        },

        /** 创建知识库 */
        async createKnowledgeBase(name: string, description?: string) {
            this.loading = true;
            try {
                const kb = await invoke<KnowledgeBase>('create_knowledge_base', {
                    name,
                    description,
                });
                this.knowledgeBases.push(kb);
                return kb;
            } catch (e) {
                this.error = String(e);
                throw e;
            } finally {
                this.loading = false;
            }
        },

        /** 删除知识库 */
        async deleteKnowledgeBase(id: string) {
            this.loading = true;
            try {
                await invoke('delete_knowledge_base', { id });
                this.knowledgeBases = this.knowledgeBases.filter(kb => kb.id !== id);
                if (this.currentKbId === id) {
                    this.currentKbId = null;
                    this.documents = [];
                }
            } catch (e) {
                this.error = String(e);
                throw e;
            } finally {
                this.loading = false;
            }
        },

        /** 选择知识库 */
        async selectKnowledgeBase(id: string) {
            this.currentKbId = id;
            await this.loadDocuments(id);
        },

        /** 加载文档列表 */
        async loadDocuments(kbId: string) {
            this.loading = true;
            try {
                this.documents = await invoke<Document[]>('get_documents', { kbId });
            } catch (e) {
                this.error = String(e);
                console.error('加载文档失败:', e);
            } finally {
                this.loading = false;
            }
        },

        /** 导入文档 */
        async importDocuments(kbId: string, paths: string[]) {
            this.loading = true;
            try {
                const docs = await invoke<Document[]>('import_documents', { kbId, paths });
                this.documents.push(...docs);
                // 更新知识库文档数
                const kb = this.knowledgeBases.find(k => k.id === kbId);
                if (kb) {
                    kb.documentCount = this.documents.length;
                }
                return docs;
            } catch (e) {
                this.error = String(e);
                throw e;
            } finally {
                this.loading = false;
            }
        },

        /** 删除文档 */
        async deleteDocument(docId: string) {
            this.loading = true;
            try {
                await invoke('delete_document', { docId });
                this.documents = this.documents.filter(doc => doc.id !== docId);
                // 更新知识库文档数
                if (this.currentKbId) {
                    const kb = this.knowledgeBases.find(k => k.id === this.currentKbId);
                    if (kb) {
                        kb.documentCount = this.documents.length;
                    }
                }
            } catch (e) {
                this.error = String(e);
                throw e;
            } finally {
                this.loading = false;
            }
        },

        /** 清除错误 */
        clearError() {
            this.error = null;
        },
    },
});

File: src-tauri/src/commands/knowledge_base.rs
//! 知识库相关的 Tauri 命令

use crate::models::{Document, KnowledgeBase};
use crate::services::knowledge_base_service::KnowledgeBaseService;

/// 创建知识库
#[tauri::command]
pub async fn create_knowledge_base(
    name: String,
    description: Option<String>,
) -> Result<KnowledgeBase, String> {
    let service = KnowledgeBaseService::new()?;
    service.create_knowledge_base(name, description)
}

/// 列出所有知识库
#[tauri::command]
pub async fn list_knowledge_bases() -> Result<Vec<KnowledgeBase>, String> {
    let service = KnowledgeBaseService::new()?;
    service.list_knowledge_bases()
}

/// 获取知识库详情
#[tauri::command]
pub async fn get_knowledge_base(id: String) -> Result<Option<KnowledgeBase>, String> {
    let service = KnowledgeBaseService::new()?;
    service.get_knowledge_base(&id)
}

/// 删除知识库
#[tauri::command]
pub async fn delete_knowledge_base(id: String) -> Result<(), String> {
    let service = KnowledgeBaseService::new()?;
    service.delete_knowledge_base(&id)
}

/// 导入文档到知识库
#[tauri::command]
pub async fn import_documents(kb_id: String, paths: Vec<String>) -> Result<Vec<Document>, String> {
    let service = KnowledgeBaseService::new()?;
    let mut documents = Vec::new();

    for path in paths {
        match service.import_document(&kb_id, &path) {
            Ok(doc) => documents.push(doc),
            Err(e) => {
                log::warn!("导入文档失败 {}: {}", path, e);
            }
        }
    }

    Ok(documents)
}

/// 获取知识库的所有文档
#[tauri::command]
pub async fn get_documents(kb_id: String) -> Result<Vec<Document>, String> {
    let service = KnowledgeBaseService::new()?;
    service.get_documents(&kb_id)
}

/// 删除文档
#[tauri::command]
pub async fn delete_document(doc_id: String) -> Result<(), String> {
    let service = KnowledgeBaseService::new()?;
    service.delete_document(&doc_id)
}
